editor/src/uuiui/inputs/number-input.tsx

```diff
@@ -389,53 +362,79 @@ export const NumberInput = React.memo<NumberInputProps>(
+        if (document.pointerLockElement === pointerOriginRef.current) {
+          document.exitPointerLock()
+        }
+        if (
+          revertChanges === 'revert-changes' &&
+          onSubmitValue != null &&
+          valueAtDragOrigin.current != null
+        ) {
+          const oldValue = cssNumber(valueAtDragOrigin.current, valueUnit)
+          onSubmitValue(oldValue, false)
+        }
+
+        setIsFauxcused(false)
+        ref.current?.focus()
+
+        setScrubThresholdPassed(false)
+        setGlobalCursor?.(null)
+      },
+      [onSubmitValue, setGlobalCursor, valueUnit],
+    )
+
+    const checkPointerLockChange = React.useCallback(() => {
+      if (document.pointerLockElement !== pointerOriginRef.current) {
+        cancelPointerLock('revert-changes')
+        scrubbingCleanupCallbacks.current.forEach((fn) => fn())
+      }
+    }, [cancelPointerLock])
+
+    const scrubOnMouseUp = React.useCallback(
       (e: MouseEvent) => {
+        scrubbingCleanupCallbacks.current.forEach((fn) => fn())
         onThresholdPassed(e, () => {
-          if (!scrubThresholdPassed.current) {
-            setScrubThresholdPassed(true)
-          }
-          setScrubValue(
-            valueUnit,
-            e.screenX,
-            e.screenY,
-            dragOriginX.current,
-            dragOriginY.current,
-            true,
-          )
+          setScrubValue(false)
         })
+        cancelPointerLock('revert-nothing')
       },
-      [setScrubValue, valueUnit],
+      [cancelPointerLock, setScrubValue],
     )
 
-    const scrubOnMouseUp = React.useCallback(
+    const scrubOnMouseMove = React.useCallback(
       (e: MouseEvent) => {
-        window.removeEventListener('mouseup', scrubOnMouseUp)
-        window.removeEventListener('mousemove', scrubOnMouseMove)
-
-        setIsFauxcused(false)
-        ref.current?.focus()
+        // Apply the movement to the accumulated delta, as the movement is
+        // relative to the last event.
+        accumulatedMouseDeltaX.current += e.movementX
 
         onThresholdPassed(e, () => {
-          setScrubValue(
-            valueUnit,
-            e.screenX,
-            e.screenY,
-            dragOriginX.current,
-            dragOriginY.current,
-            false,
-          )
+          if (!scrubThresholdPassed.current) {
+            setScrubThresholdPassed(true)
+            if (pointerOriginRef.current != null) {
+              pointerOriginRef.current.requestPointerLock()
+              scrubbingCleanupCallbacks.current.push(() => {
+                window.removeEventListener('mouseup', scrubOnMouseUp)
+              })
+              scrubbingCleanupCallbacks.current.push(() => {
+                window.removeEventListener('mousemove', scrubOnMouseMove)
+              })
+              scrubbingCleanupCallbacks.current.push(() => {
+                document.removeEventListener('pointerlockchange', checkPointerLockChange, true)
+              })
```

Comment:
Is there a reason to push these cleanups to a ref of an array like this? It seems to me that in both cleanup sites (`checkPointerLockChange` and `scrubOnMouseUp`) we could just call a single function that removes these 3 event listeners regardless of the circumstances

Reply:
I referenced this in the description, the issue was that I kept ending up with circular references between the callbacks, as their dependencies need to be declared before the callbacks which use them. This means that the function which triggers the cleanup doesn't depend on the event listeners, which snips a convenient gap in the circle.

Reply:
Ohhhh.... ew.... ok, can you add that in a comment here please too

Reply:
Fixed in 8e2c64b2b.